job-server/src/spark.jobserver/util/SparkMasterProvider.scala

```diff
@@ -0,0 +1,54 @@
+package spark.jobserver.util
+
+import scala.reflect.runtime.{universe => ru}
+import com.typesafe.config.{ConfigException, Config}
+import org.slf4j.LoggerFactory
+
+
+trait SparkMasterProvider {
+
+  /**
+   * Implementing classes will determine what the appropriate SparkMaster is and
+   * return it
+   * @return A Spark Master Address
+   */
+  def getSparkMaster(config: Config): String
+
+}
+
+object SparkMasterProvider {
+  val logger = LoggerFactory.getLogger(getClass)
+  val SparkMasterProperty = "spark.master.provider"
+
+  /**
+   * Will look for an Object with the name provided in the Config file and return it
+   * or the DefaultSparkMasterProvider if no spark.master.provider was specified
+   * @param config SparkJobserver Config
+   * @return A SparkMasterProvider
+   */
+  def fromConfig(config: Config): SparkMasterProvider = {
+
+    try {
+      val sparkMasterObjectName = config.getString(SparkMasterProperty)
+      println(s"Using $sparkMasterObjectName to determine Spark Master")
+      logger.info(s"Using $sparkMasterObjectName to determine Spark Master")
+      val m = ru.runtimeMirror(getClass.getClassLoader)
+      val module = m.staticModule(sparkMasterObjectName)
+      val sparkMasterProviderObject = m.reflectModule(module).instance
+        .asInstanceOf[SparkMasterProvider]
+      sparkMasterProviderObject
+    } catch {
+      case me: ConfigException => DefaultSparkMasterProvider
```

Comment:
You could get rid of this line by providing a default in `application.conf`.... I think....